Mathlib/LinearAlgebra/Matrix/SVD/Defs.lean

```diff
@@ -0,0 +1,479 @@
+scaling by the singular values. -/
+noncomputable def svdU₁ (A : Matrix (Fin M) (Fin N) 𝕂) : Matrix (Fin M) (Fin A.rank) 𝕂 :=
+  A  *  A.svdV₁  *  (A.svdσ.map (algebraMap ℝ 𝕂))⁻¹
+
+/-- The left eigenvectors of a matrix A corresponding to its non-zero eigenvaules, obtained directly
+from the eigendecomposition of the AAᴴ matrix. These do NOT share the same ordering as `svdU₁`. -/
+noncomputable def svdU₁' (A : Matrix (Fin M) (Fin N) 𝕂) : Matrix (Fin M) (Fin (A.rank)) 𝕂 :=
+  ((reindex (Equiv.refl (Fin M)) (eigenRowEquiv A))
+    (isHermitian_mul_conjTranspose_self A).eigenvectorMatrix).toColumns₁
+
+/-- The left eigenvectors of a matrix A corresponding to its zero eigenvaules, obtained directly
+from the eigendecomposition of the AAᴴ matrix. The order of these eigenvectors is not relevatn as
+they are multiplied by zero anyway. Hence we do not have `svdU₂` -/
+noncomputable def svdU₂ (A : Matrix (Fin M) (Fin N) 𝕂) : Matrix (Fin M) (Fin (M - A.rank)) 𝕂 :=
+  ((reindex (Equiv.refl (Fin M)) (eigenRowEquiv A))
+    (isHermitian_mul_conjTranspose_self A).eigenvectorMatrix).toColumns₂
+
+/-- Concatenation of the left eigenvectors U₁ and U₂ into one eigenvector matrix. This is a unitary
+matrix. -/
+noncomputable def svdU (A : Matrix (Fin M) (Fin N) 𝕂) :
+    Matrix (Fin M) (Fin (A.rank) ⊕ Fin (M - A.rank)) 𝕂 := fromColumns A.svdU₁ A.svdU₂
+
+/-- Concatenation of the right eigenvectors V₁ and V₂ into one eigenvector matrix. This is a unitary
+matrix. -/
+noncomputable def svdV (A : Matrix (Fin M) (Fin N) 𝕂) :
+    Matrix (Fin N) (Fin (A.rank) ⊕ Fin (N - A.rank)) 𝕂 := fromColumns A.svdV₁ A.svdV₂
+
+/-- Given a matrix A of size m × n: `svdS` is a matrix of the same dimensions but partitioned into
+four blocks such that S₁₁ contains the non-zero singular values, S₁₂, S₂₁ and S₂₂ are zeros. The S₁₁
+block is the diagonal matrix `svdσ` above. -/
+noncomputable def svdS (A : Matrix (Fin M) (Fin N) 𝕂) :
+    Matrix ((Fin A.rank) ⊕ (Fin (N - A.rank))) ((Fin A.rank) ⊕ (Fin (N - A.rank))) ℝ :=
+  (reindex (eigenColumnEquiv A) (eigenColumnEquiv A))
+    (diagonal (isHermitian_transpose_mul_self A).eigenvalues)
+
+/-- Given a matrix A of size m × n: `svdS'` is a matrix of the same dimensions but partitioned into
+four blocks such that S₁₁ contains the non-zero singular values, S₁₂, S₂₁ and S₂₂ are zeros. The S₁₁
+block is the diagonal matrix `svdσ` above but permuted by the same arbitrary order relating svdU₁
+and svdU₁' above . -/
+noncomputable def svdS' (A : Matrix (Fin M) (Fin N) 𝕂) :
+    Matrix ((Fin A.rank) ⊕ (Fin (M - A.rank))) ((Fin A.rank) ⊕ (Fin (M - A.rank))) ℝ :=
+  (reindex (eigenRowEquiv A) (eigenRowEquiv A))
+    (diagonal (isHermitian_mul_conjTranspose_self A).eigenvalues)
+
+lemma U_columns' (A : Matrix (Fin M) (Fin N) 𝕂) :
+  ((reindex (Equiv.refl (Fin M)) (eigenRowEquiv A))
+    (isHermitian_mul_conjTranspose_self A).eigenvectorMatrix) = fromColumns A.svdU₁' A.svdU₂ := by
```

Comment:
My experience with mathlib names so far is that they are difficult to remember (sometimes even for the people who wrote the lemma them unless they are in the same "frame" of mind they where in at time of writing). One possible reason for this is because I do not understand how it works.

I made this particular change. but how do I arrive at it, to save future reviewers some time?
The naming convention document here https://leanprover-community.github.io/contribute/naming.html
How do I reach from the theorem to the name you came up with by following that documnet (or others if there are any)?

Reply:
The important part for this question is the section on [Identifiers and theorem names](https://leanprover-community.github.io/contribute/naming.html#identifiers-and-theorem-names), especially the part that starts

> Often the name of theorem simply describes the conclusion:

The algorithm described there for coming up with a name works essentially as follows.

First write down the names in the conclusion, joined by underscores.
> `reindex_refl_Fin_eigenRowEquiv_isHermitian_mul_conjTranspose_self_eigenvectorMatrix_eq_fromColumns_svdU₁'_svdU₂`
If this name is clear and short enough, you're done! Otherwise, you can:
 * shorten by eliminating parts that wouldn't be confusing to leave out. For example `eigenRowEquiv` is about the eigenvalues of `A⬝Aᴴ` so we can leave out the `isHermitian_mul_conjTranspose_self` part below. Similarly if we don't do anything with the columns we might as well leave out `refl_Fin`:
> `reindex_eigenRowEquiv_eigenvectorMatrix_eq_fromColumns_svdU₁'_svdU₂`
 * shorten by eliminating the right hand side of the equality (as a rule of thumb the right hand side is easy to guess: generally you want one way to write things, so `B` should be obvious from `A = B`)
> `reindex_eigenRowEquiv_eigenvectorMatrix`
 * expand by adding hypotheses, separated by `_of_`. Imagine if this result required `A` to be nonzero, then we'd write `reindex_eigenRowEquiv_eigenvectorMatrix_of_ne_zero`.

Following the naming convention really is important because experienced Mathlib users can find results in a different part of the library by just writing what they think this result should be called. It cost much less to guess a name than to switch to your browser and type a query into a search engine like Loogle.

In the end it's going to be a matter of taste which parts to emphasize and which to leave out; there is no one right answer. But following the naming convention is very helpful.